### PR TITLE
[SYCL] annotated_ptr API fixes part 3

### PR DESCRIPTION
**Pull request by rolandschulz**
Deprecate implicit constructor by marking constructor explicit and adding new deprecated implicit constructor.

This is separate because it won't be included in 2024.0.

### CODE REVIEW

**Comment by rolandschulz** [comment]
Don't merge until part 2 (https://github.com/intel/llvm/pull/10928) is merged.

**Comment by bader** [comment]
> Don't merge until part 2 (#10928) is merged.

I converted this PR to draft, which can't be merged.
Once #10928 is merged, please, push "ready for review" button.

**Review by Pennycook — APPROVED** [review]

**Review by broxigarchen — APPROVED** [review]

File: sycl/include/sycl/ext/oneapi/experimental/annotated_ptr/annotated_ptr.hpp
Comment (GarveyJoe) [inline sycl/include/sycl/ext/oneapi/experimental/annotated_ptr/annotated_ptr.hpp:268]: Given that we've decided not to document annotated_ptr for 2024.0, I think this part can be left out.  It will be present in the header files for 2024.0 but I don't think we have a duty to keep undocumented features working.

Comment (tiwaria1) [reply sycl/include/sycl/ext/oneapi/experimental/annotated_ptr/annotated_ptr.hpp:268]: Agreed, Roland please remove the deprecation warning.

**Review by tiwaria1 — CHANGES_REQUESTED** [review]

**Review by tiwaria1 — APPROVED** [review]

**Review by aelovikov-intel — APPROVED** [review]

**Comment by aelovikov-intel** [comment]
@rolandschulz do you want me to merge this? Or are we waiting for somebody else to take a look?

**Comment by rolandschulz** [comment]
I think this can be merged

**Pull request merged**
